### PR TITLE
Merge 3.20.x to main

### PR DESCRIPTION
**Pull request by haberman**
@JasonLunn please take a close look at the JRuby stuff, there were some merge conflicts that I mostly resolved in favor of main (it looks like there were some formatting fixes perhaps?).

### CODE REVIEW

**Comment by JasonLunn** [comment]
@haberman - 100% of conflicts should be resolved in favor of `main`; what was brought into the `3.20.x` branch (via #9714) were functional fixes, but #9690 and #9697 were not included in what was pulled into the release candidate branch. Looking at the Ruby failures above, the merge conflicts appear to have left `RubyMessage` in a syntactically invalid state.

File: ruby/src/main/java/com/google/protobuf/jruby/RubyMessage.java [hunk 1]
```diff
@@ -860,6 +860,7 @@ protected DynamicMessage build(ThreadContext context, int depth, int recursionLi
               this.builder.getDescriptorForType().findFieldByName(TYPE);
           if (typeBytesSymbol.equals(fields.get(enumFieldDescriptorForType))) {
             isDefaultStringForBytes = true;
+                    }
```
Comment (JasonLunn) [inline ruby/src/main/java/com/google/protobuf/jruby/RubyMessage.java:863]: Specifically, this additional `}` seems to be the culprit

File: ruby/src/main/java/com/google/protobuf/jruby/RubyDescriptor.java [hunk 2]
```diff
@@ -215,8 +215,8 @@ public IRubyObject allocate(Ruby runtime, RubyClass klazz) {
     klass.include(new IRubyObject[] {messageExts});
     klass.instance_variable_set(runtime.newString(Utils.DESCRIPTOR_INSTANCE_VAR), this);
     klass.defineAnnotatedMethods(RubyMessage.class);
-    // Workaround for https://github.com/jruby/jruby/issues/7154
-    klass.searchMethod("respond_to?").setIsBuiltin(false);
+        // Workaround for https://github.com/jruby/jruby/issues/7154
```
Comment (JasonLunn) [inline ruby/src/main/java/com/google/protobuf/jruby/RubyDescriptor.java:218]: It isn't clear that this indentation change is intended.

**Review by acozzette — APPROVED** [review]

File: ruby/src/main/java/com/google/protobuf/jruby/RubyMessage.java (hunk 1 above)
Comment (haberman) [reply ruby/src/main/java/com/google/protobuf/jruby/RubyMessage.java]: Done.

File: ruby/src/main/java/com/google/protobuf/jruby/RubyDescriptor.java (hunk 2 above)
Comment (haberman) [reply ruby/src/main/java/com/google/protobuf/jruby/RubyDescriptor.java]: Reverted.

**Pull request merged**